# ML-03 — Frame Your Lane as an ML Task

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane as an ML task (type)

*Classification, clustering, ranking, or scoring — which one, and why?*

My chosen lane is Content Refresh / Opportunity Scoring.

The main ML task is scoring because the system will assign each content page a refresh opportunity score. A higher score means that the page should receive more attention and may need to be updated sooner.

The scores can also be used to rank pages, helping the content team decide which pages to review first.

## 2. Target or proxy

*What would you predict? Where does that label come from — observed outcome or a defined rule?*

The target is whether a content page is showing a decline in performance.

Because the dataset does not contain the real result of refreshing each page, I will use performance decline as a proxy for refresh need. The proxy label can be created from observed historical performance, such as comparing recent impressions with impressions from an earlier period.

The model will predict the probability of decline, and this probability can contribute to the refresh opportunity score.

## 3. Success metric

*One metric you can defend. What number means 'good'?*

The main success metric is Precision@50.

Precision@50 measures how many of the 50 highest-scored pages are actually
showing a decline in performance. This metric matches the real action because
the content team has limited time and will review only the highest-priority
pages.

I would consider a Precision@50 of at least 0.70 good. This means that at least
35 of the top 50 recommended pages are correctly identified as declining.

## 4. The unit of analysis, as a real dataframe

*Load your lane's slice and show it: one row = one what?*

In [6]:
import pandas as pd
df = pd.read_csv("../../data/raw/content_refresh_anonymized.csv")
print("Dataset shape:", df.shape)
print("Column names:")
print(df.columns.tolist())
df.head()

lane_df = df[
    [
        "content_id",
        "search_volume",
        "impressions_last_30d",
        "impressions_prev_30d",
        "days_since_last_update",
        "avg_position",
        "engagement_rate",
        "trend_direction",
        "trend_pct"
    ]
].copy()
print("Unit of analysis: one row represents one content page.")
lane_df.head()
lane_df["decline_target"] = (
    lane_df["trend_direction"] == "down"
).astype(int)
lane_df[
    [
        "content_id",
        "impressions_prev_30d",
        "impressions_last_30d",
        "trend_direction",
        "trend_pct",
        "decline_target"
    ]
].head(10)

Dataset shape: (30000, 44)
Column names:
['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']
Unit of analysis: one row represents one content page.


,content_id,impressions_prev_30d,impressions_last_30d,trend_direction,trend_pct,decline_target
0,content_304f48230142,987,578,down,-41.4,1
1,content_a1fb4e703a9e,5915,2501,down,-57.7,1
2,content_9aa793d4d895,6089,2382,down,-60.9,1
3,content_331d6c4de07b,4206,3626,stable,-13.8,0
4,content_d99b7a2d90ca,6452,4211,down,-34.7,1
5,content_d4084a4bc775,1009,617,down,-38.9,1
6,content_9a34b442b552,13,1,down,-92.3,1
7,content_a63219c6e95a,632,636,stable,0.6,0
8,content_5e6c160719bc,13828,5696,down,-58.8,1
9,content_c27558df2b0c,356,252,down,-29.2,1


## 5. Why ML beats a fixed rule here

*What makes the pattern too messy for an if-statement?*

A fixed rule based only on trend percentage would not be enough because a large percentage decline does not always mean a large business impact.

For example, a page that falls from 13 impressions to 1 impression has a very large percentage decline, but it may not be as important as a page that falls from 13,828 impressions to 5,696 impressions.

Machine learning can combine several signals such as search volume, current impressions, average position, engagement rate, content age, and days since the last update. This allows the system to identify more valuable refresh opportunities instead of depending on only one rule.

The output supports the content team by ranking pages and helping them decide which content should be reviewed and refreshed first.

## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.